openhtf/__init__.py

```diff
@@ -159,21 +164,27 @@ def add_output_callbacks(self, *callbacks):
     self._test_options.output_callbacks.extend(callbacks)
 
   def configure(self, **kwargs):
-    """Update test-wide configuration options.
-
-    Valid kwargs:
-      output_callbacks: List of output callbacks to run, typically it's better
-          to use add_output_callbacks(), but you can pass [] here to reset them.
-      teardown_function: Function to run at teardown.  We pass the same
-          arguments to it as a phase.
-    """
+    """Update test-wide configuration options. See TestOptions for docs."""
     # These internally ensure they are safe to call multiple times with no weird
     # side effects.
     create_arg_parser(add_help=True).parse_known_args()
     logs.setup_logger()
     for key, value in kwargs.iteritems():
       setattr(self._test_options, key, value)
 
+    # 'capture_source' takes effect here.
+    if self._test_options.capture_source:
+      # First, we copy the phases with the real CodeInfo for them.
+      phases = [
+        mutablerecords.CopyRecord(
+            phase, code_info=test_record.CodeInfo.for_function(phase.func))
+        for phase in self._test_desc.phases]
+
+      # Then we replace the TestDescriptor with one that stores the test
+      # module's CodeInfo as well as our newly copied phases.
+      self._test_desc = self._test_desc._replace(
+          code_info=self.code_info, phases=phases)
```

Comment:
`namedtuple` exports it as `_replace` to not conflict with potential attribute names (since they're enforced as not starting with `_`). We could separate out the code info into separate objects instead of inside the `PhaseDescriptor`/`TestDescriptor`?

We have some other options:
* Make `capture_source` a kwarg to `Test` so we know whether to grab the code info as early as that.
* Store a `TestDescriptor` in an unconfigured state, maybe in `self._initial_test_desc` then only after calling `configure()` they get access to it. `descriptor` is documented as not changing across `execute()` calls, not `configure()` calls.
* Separate the `CodeInfo` to be stored in a separate set of objects than from `TestDescriptor`, though it does make sense to put it in the 'static test information' :/